### PR TITLE
Fix/nex 778/prevent override persistence

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
Installation fix to work with a spanner. 

https://oat-sa.atlassian.net/browse/NEX-778

### CODE REVIEW

File: scripts/update/Updater.php [hunk 1]
```diff
@@ -1343,8 +1343,8 @@ public function update($initialVersion)
                 )
             );
 
-            $this->setVersion('41.2.0');
+            $this->setVersion('41.2.1');
```
Comment (mike-ionut-mihai-sandu-tao) [inline scripts/update/Updater.php]: why ?

File: .Jenkinsfile [hunk 2]
```diff
@@ -53,6 +53,15 @@ registry.service.consul:4444/tao/dependency-resolver oat:dependencies:resolve --
                         label: 'Add phpunit',
                         script: 'composer require phpunit/phpunit:^8.5'
                     )
+                    sh(
```
Comment (mike-ionut-mihai-sandu-tao) [inline .Jenkinsfile]: this should be in a different PR

File: install/class.Setup.php [hunk 3]
```diff
@@ -263,6 +213,16 @@ public function __invoke($params)
 
         $serviceManager = $installator->getServiceManager();
 
+        if (!isset($parameters['configuration']['generis']['persistences'])) {
+            throw new InvalidArgumentException('Your config should have a \'persistence\' key under \'generis\'');
```
Comment (mike-ionut-mihai-sandu-tao) [inline install/class.Setup.php]: is it a InvalidArgumentException or runtimeException ?

File: install/class.Setup.php [hunk 4]
```diff
@@ -263,6 +213,16 @@ public function __invoke($params)
...
+        }
+        $persistences = $parameters['configuration']['generis']['persistences'];
+        if (isset($persistences['default'])) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline install/class.Setup.php]: this can be set to '' and isset will return true
```
$persistences['default'] = '';

 if (isset($persistences['default'])) {
     echo  'breaking somthing';
 }
```
```suggestion
         if (!empty($persistences['default'])) {
```

File: install/class.Setup.php [hunk 5]
```diff
@@ -327,4 +281,59 @@ public function __invoke($params)
...
+     * and then back into a persistence configuration to ensure backwards compatibility
+     * with the previous process
+     * @param array $persistences
+     * @return array
+     */
+    private function wrapPersistenceConfig($persistences)
+    {
+        $defaultPersistenceConfig = $persistences['default'];
+        $options = [];
+        if (isset($defaultPersistenceConfig['connection'])) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline install/class.Setup.php]: a lot of complexity here
switch case could help us to make it a little bit more structured?

File: scripts/update/Updater.php (hunk 1 above)
Comment (jbout) [reply scripts/update/Updater.php]: Looks like a bad merge

File: install/class.Setup.php (hunk 3 above)
Comment (jbout) [reply install/class.Setup.php]: Unchanged move from line 150, all other checks also throw InvalidArgumentException, would not recommend changing the behaviour in the context of this PR.

File: install/class.Setup.php (hunk 4 above)
Comment (jbout) [reply install/class.Setup.php]: This is not a validation but a test to see if persistences are defined as a service or as direct configuration parameters:
if $persistences['default'] is set => direct persistence configuration parameters
if $persistences['type'] is set => full service configuration
if neither is set => PROBLEM

else there would be many more tests required: Is is an array, does it contain a key db_driver and so on

File: install/class.Setup.php [hunk 6]
```diff
@@ -327,4 +281,100 @@ public function __invoke($params)
...
     }
+
+    /**
+     * Transforms the seed persistence configuration into command line parameters
+     * and then back into a persistence configuration to ensure backwards compatibility
+     * with the previous process
+     * @param array $persistences
+     * @return array
+     */
+    private function wrapPersistenceConfig($persistences)
```
Comment (jbout) [inline install/class.Setup.php]: While I see how this config was difficult to read, this split makes it even harder for me to understand what is happening:
* we generate the command line parameters ($options) based on the seed config of the default persistence
* we then generate a new persistence config based on the command line parameters
* we replace the default persistence config in the seed, by the newly generated one

File: install/class.Setup.php [hunk 7]
```diff
@@ -327,4 +281,72 @@ public function __invoke($params)
...
+    {
+        $seedConfiguration = $this->getCommandLineParameters($persistences['default']);
```
Comment (jbout) [inline install/class.Setup.php]: Much clearer now.
However calling the command line parameter array "seedConfiguration" is still misleading, as the seed config actually gets generated in line 297

Comment (bartlomiejmarszal) [reply install/class.Setup.php]: How would u called?
```suggestion
        $configuration = $this->getCommandLineParameters($persistences['default']);
```

Comment (bartlomiejmarszal) [reply install/class.Setup.php]:
```suggestion
        $parsedConfiguration = $this->getCommandLineParameters($persistences['default']);
```

**Comment by mike-ionut-mihai-sandu-tao** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
